Add application-defined Reference space transforms ("Teleportation")

toji opened the pull request:
Follow up to #440, second part of a solution to #431. This PR adds the ability for developers to specify a transform to any `XRReferenceSpace` that will be applied to any sensor-controlled poses, effectively giving developers a simple method of "teleporting" the user around a scene. (That's a bit over-specific, since this also allows for things like touch/mouse rotation of magic window content and continuous stick motion if you have the stomach for it.)

Technically this is something that any dev can do themselves, but it means ensuring that you're consistently applying the transform to all your inputs, manually reversing it out for things like hit testing, and inverting, applying, and re-inverting in order to apply to view matrices. Thus handling it in the API is a significant convenience (that developers can still ignore if they are so inclined). It's also something that can be handled entirely at the API implementation level and won't need specialized logic for every native backend, so realistically the code will need to be written ~3 times and then everyone benefits forever. 

As mentioned in #440, the fact that this uses an `XRRigidTransform` rather than a raw `Float32Array` matrix is important because it allows us to impose constraints/validation on the types of transforms being done. No weird shearing, for example.

thetuvix commented on spatial-tracking-explainer.md:
````diff
@@ -177,6 +177,33 @@ function onSessionStarted(session) {
+    xrSession.requestAnimationFrame(onDrawFrame);
+  });
+}
+```
````
If this static scene start offset was the only thing in play, we'd likely just suggest that scene authors shift their content around so that the static starting location is at (0, 0, 0).  The motivating bit comes in later on, when the user teleports within the scene.

It may be worth also showing a `function teleport(transform)` that later mutates the transform on `xrReferenceSpace`, reading it back, applying the offset and then setting it again.

toji replied:
Good idea. Done!

NellWaliczek replied:
I'm actually not sure the previous example is super helpful.  The teleportation one below I'm all for!

thetuvix commented on spatial-tracking-explainer.md:
```diff
@@ -177,6 +177,33 @@ function onSessionStarted(session) {
+
+Developers can specify application-specific transforms by setting the `baseTransform` attribute of any `XRReferenceSpace`. When the `transform` is not null, any values queried using the `XRReferenceSpace` will be offset by the `translation` and `orientation` the `baseTransform` describes. The `XRReferenceSpace`'s `baseTransform` can be updated at any time and will immediately take effect, meaning ...
+
+When interpreting an `XRRigidTransform` the `orientation` is applied prior to the `translation`. This means that, for example, a transform that indicates a quarter rotation to the right and a 1 meter translation along -Z would place a transformed object at `[0, 0, -1]` facing to the right.
```
This is yet another benefit to `position`/`orientation` naming - developers have a natural instinct around what it means for an entity to be positioned and oriented within a coordinate system and thus would be unlikely to get this backwards.

thetuvix replied:
Note that the explanation you provide here for how to interpret the two attributes describes an object that is placed (i.e. positioned) at (0, 0, -1) with a given orientation.  If the attributes had those two names directly, that would likely be immediately obvious to developers.

toji replied:
#440 has been updated to use `position/orientation` now. Thanks again for the suggestion.

blairmacintyre commented on spatial-tracking-explainer.md:
```diff
@@ -364,7 +391,9 @@ dictionary XRReferenceSpaceOptions {
 
-[SecureContext, Exposed=Window] interface XRReferenceSpace : XRSpace {
+[SecureContext, Exposed=Window] interface XRReferenceSpace : XRSpace {  
+  attribute XRRigidTransform? baseTransform;
```
I can't recall who suggested offset, but I liked that too.  `originOffset` 

I do not see offset as implying translation only, but I can see how some might, though.

I'm not a fan of "base".

Or, `originOffset`?

toji replied:
I suspect you intended to not suggest `originOffset` twice, so now I'm curious what the other suggestion was supposed to be? 😉 

In general, though, I'm seeing a lot of support for the term `origin` and more support than expected for some variant of `offset`, which likely means that my concern about it's implied meaning is overblown. Going to move for the moment to `originTransform` since it gets away from the `base` term that doesn't seem too popular and I've now heard it suggested from a couple of different sources (thanks @darktears!) but I could be persuaded to go with `originOffset` if that has more support.

darktears replied:
As it was highlighted in IRC yesterday (and thus my suggestion in the first place) origin is consistent with the web platform, see https://developer.mozilla.org/en-US/docs/Web/CSS/transform-origin

toji replied:
I missed that in IRC yesterday, thanks for repeating it here! I definitely think consistency with the rest of the web platform is a positive thing!

thetuvix replied:
I do somewhat prefer `originOffset` to `originTransform`, as it more strongly suggests that there is a natural origin that is meaningful in some way, but which you can offset relative to that natural location if you need to, e.g. to accomplish teleportation.

Most other attributes in the API named `transform` or `...Transform` can be though of fairly simply as "the absolute position and orientation of the thing within the space", which is not really what this relative offset is, and so perhaps it's good that the attribute name is a little different.

NellWaliczek commented on spatial-tracking-explainer.md:
````diff
@@ -177,6 +177,45 @@ function onSessionStarted(session) {
 ```
 
+### Application-supplied transforms
+Frequently developers will want to provide an additional, artificial transform on top of the user's tracked motion to allow the user to navigate larger scenes than their tracking systems or physical space allows. This effect is traditionally accomplished by mathematically combining the API-provided transform with the desired additional application transforms. It is useful to allow the API do hand ...
````
nit: navigate larger scenes  -> navigate larger virtual scenes

NellWaliczek replied:
nit: API do handle it -> API to handle it

NellWaliczek replied:
Nit: this is a bit confusing sounding to me...

>It is useful to allow the API do handle it, however, however, because it ensures that all tracked values are transformed consistently, allowing the user's viewpoint and inputs to stay in sync.

How about this instead:

Nit: this is a bit confusing sounding to me...

> WebXR offers developers a simplification to ensure that all tracked values are transformed consistently, allowing the user's viewpoint and inputs to stay in sync.

NellWaliczek commented on spatial-tracking-explainer.md:
```diff
@@ -364,7 +403,9 @@ dictionary XRReferenceSpaceOptions {
 
-[SecureContext, Exposed=Window] interface XRReferenceSpace : XRSpace {
+[SecureContext, Exposed=Window] interface XRReferenceSpace : XRSpace {  
+  attribute XRRigidTransform? originTransform;
```
Is there usefulness in this being nullable?  Or can we get away with just making it be identity by default?

toji replied:
Seems reasonable to me!

NellWaliczek commented on spatial-tracking-explainer.md:
````diff
@@ -177,6 +177,45 @@ function onSessionStarted(session) {
+}
+```
+
+Another common use case for this attribute would be for a "teleportation" mechanic, where the user "jumps" to a new point in the virtual scene, after which the selected point is treated as the new origin which all tracked motion is relative to.
````
nit: the new origin -> the new virtual origin

NellWaliczek commented on spatial-tracking-explainer.md:
````diff
@@ -177,6 +177,45 @@ function onSessionStarted(session) {
 ```
 
+### Application-supplied transforms
+Frequently developers will want to provide an additional, artificial transform on top of the user's tracked motion to allow the user to navigate larger virtual scenes than their tracking systems or physical space allows. This effect is traditionally accomplished by mathematically combining the API-provided transform with the desired additional application transforms. WebXR offers developers a sim ...
````
nit: maybe give an example of two types of tracked values in the last sentence

NellWaliczek commented on spatial-tracking-explainer.md:
```diff
@@ -177,6 +177,45 @@ function onSessionStarted(session) {
+### Application-supplied transforms
+Frequently developers will want to provide an additional, artificial transform on top of the user's tracked motion to allow the user to navigate larger virtual scenes than their tracking systems or physical space allows. This effect is traditionally accomplished by mathematically combining the API-provided transform with the desired additional application transforms. WebXR offers developers a sim ...
+
+Developers can specify application-specific transforms by setting the `originTransform` attribute of any `XRReferenceSpace`. When the `transform` is not null, any values queried using the `XRReferenceSpace` will be offset by the `position` and `orientation` the `originTransform` describes. The `XRReferenceSpace`'s `originTransform` can be updated at any time and will immediately take effect, mean ...
```
>  When the `transform` is not null

This sentence needs to be fixed now that this matrix isn't nullable

NellWaliczek replied:
> meaning that any new values that are queried with the `XRReferenceSpace` will take into account the new `originTransform`.

perhaps could be
> meaning that any subsequent poses queried with the `XRReferenceSpace` will take into account the new `originTransform`.

toji commented:
Addressed the latest round of feedback. Biggest change is that I have switched from `originTransform` to `originOffset` due to some of the well reasoned arguments that have been presented in it's favor, as well as simple numbers. (Nell, Alex, and Blair have voiced a preference for `offset`, while Alexis and myself have been the ones favoring `transform`)
